core/trino-main/src/main/java/io/trino/exchange/LazyExchangeDataSource.java

```diff
@@ -102,7 +101,9 @@ public ListenableFuture<Void> isBlocked()
             return initializationFuture;
         }
         ExchangeDataSource dataSource = delegate.get();
-        checkState(dataSource != null, "dataSource is expected to be initialized");
+        if (dataSource == null) {
```

Comment:
The `Allow isBlocked after close in ExchangeDataSource` commit should be moved to be before `Support consuming query results from a spooling exchange`. Otherwise you are introducing a bug-window.